crates/sui-core/src/transaction_driver/unit_tests/effects_certifier_tests.rs

```diff
@@ -0,0 +1,903 @@
+    messages_checkpoint::{
+        CheckpointRequest, CheckpointRequestV2, CheckpointResponse, CheckpointResponseV2,
+    },
+    messages_consensus::ConsensusPosition,
+    messages_grpc::{
+        HandleCertificateRequestV3, HandleCertificateResponseV2, HandleCertificateResponseV3,
+        HandleSoftBundleCertificatesRequestV3, HandleSoftBundleCertificatesResponseV3,
+        HandleTransactionResponse, ObjectInfoRequest, ObjectInfoResponse, RawSubmitTxRequest,
+        RawSubmitTxResponse, RawWaitForEffectsRequest, RawWaitForEffectsResponse,
+        SystemStateRequest, TransactionInfoRequest, TransactionInfoResponse,
+    },
+    quorum_driver_types::EffectsFinalityInfo,
+    sui_system_state::SuiSystemState,
+    transaction::{CertifiedTransaction, Transaction},
+};
+use tokio::time::{sleep, Duration};
+
+// Mock AuthorityAPI for testing
+#[derive(Clone)]
+struct MockAuthority {
+    _name: AuthorityName,
+    ack_responses: Arc<StdMutex<HashMap<TransactionDigest, WaitForEffectsResponse>>>,
+    full_responses: Arc<StdMutex<HashMap<TransactionDigest, WaitForEffectsResponse>>>,
+    delay: Option<Duration>,
+    should_timeout: bool,
+}
+
+impl MockAuthority {
+    fn new(name: AuthorityName) -> Self {
+        Self {
+            _name: name,
+            ack_responses: Arc::new(StdMutex::new(HashMap::new())),
+            full_responses: Arc::new(StdMutex::new(HashMap::new())),
+            delay: None,
+            should_timeout: false,
+        }
+    }
+
+    fn set_ack_response(&self, tx_digest: TransactionDigest, response: WaitForEffectsResponse) {
+        self.ack_responses
+            .lock()
+            .unwrap()
+            .insert(tx_digest, response);
+    }
+
+    fn set_full_response(&self, tx_digest: TransactionDigest, response: WaitForEffectsResponse) {
+        self.full_responses
+            .lock()
+            .unwrap()
+            .insert(tx_digest, response);
+    }
+}
+
+#[async_trait]
+impl AuthorityAPI for MockAuthority {
+    async fn wait_for_effects(
+        &self,
+        request: RawWaitForEffectsRequest,
+        _client_addr: Option<SocketAddr>,
+    ) -> Result<RawWaitForEffectsResponse, SuiError> {
+        if self.should_timeout {
+            sleep(Duration::from_secs(10)).await;
+        }
+
+        if let Some(delay) = self.delay {
+            sleep(delay).await;
+        }
+
+        let wait_request: WaitForEffectsRequest = request.try_into()?;
+
+        // Choose the right response based on include_details flag
+        let responses = if wait_request.include_details {
+            &self.full_responses
+        } else {
+            &self.ack_responses
+        };
+
+        let responses = responses.lock().unwrap();
+
+        if let Some(response) = responses.get(&wait_request.transaction_digest) {
+            let raw_response = RawWaitForEffectsResponse::try_from(response.clone())
+                .map_err(|_| SuiError::Unknown("Conversion failed".to_string()))?;
```

Comment:
To simulate the actual handler logic (notify read on effects that will never arrive), should this just wait indefinitely instead?

Also, I think `Unknown` "should" be used for unknown data / state. Usually xxNotFound errors are suitable for lookup failures, if that's the intention here.